// @aws-cdk/integ-tests-alpha/lib/manifest-synthesizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegManifestSynthesizer = void 0;
const cloud_assembly_schema_1 = require("aws-cdk-lib/cloud-assembly-schema");
const manifest_writer_1 = require("./manifest-writer");
const emptyManifest = {
    version: '',
    testCases: {},
};
class IntegManifestSynthesizer {
    constructor(testCases, enableLookups) {
        this.testCases = testCases;
        this.enableLookups = enableLookups;
    }
    synthesize(session) {
        const manifest = {
            enableLookups: this.enableLookups,
            ...this.testCases
                .map(tc => tc.manifest)
                .reduce(mergeManifests, emptyManifest),
        };
        const snapshotDir = session.assembly.outdir;
        manifest_writer_1.IntegManifestWriter.write(manifest, snapshotDir);
    }
}
exports.IntegManifestSynthesizer = IntegManifestSynthesizer;
function mergeManifests(m1, m2) {
    return {
        version: cloud_assembly_schema_1.Manifest.version(),
        testCases: { ...m1.testCases, ...m2.testCases },
    };
}
