// @aws-cdk/integ-tests-alpha/lib/assertions/providers/lambda-handler/http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpHandler = void 0;
/* eslint-disable no-console */
// eslint-disable-next-line import/no-extraneous-dependencies
const node_fetch_1 = require("node-fetch");
// TODO: can use builtin fetch on node18
const aws_custom_resource_sdk_adapter_1 = require("@aws-cdk/aws-custom-resource-sdk-adapter");
const base_1 = require("./base");
const utils_1 = require("./utils");
class HttpHandler extends base_1.CustomResourceHandler {
    async processEvent(request) {
        console.log('request', request);
        const response = await (0, node_fetch_1.default)(request.parameters.url, request.parameters.fetchOptions);
        const result = {
            ok: response.ok,
            status: response.status,
            statusText: response.statusText,
            headers: response.headers.raw(),
        };
        result.body = await response.text();
        try {
            result.body = JSON.parse(result.body);
        }
        catch (e) {
            // Okay
        }
        let resp;
        if (request.flattenResponse === 'true') {
            // Flatten and explode JSON fields
            resp = (0, aws_custom_resource_sdk_adapter_1.flatten)((0, utils_1.deepParseJson)({ apiCallResponse: result }));
        }
        else {
            // Otherwise just return the response as-is, without exploding JSON fields
            resp = { apiCallResponse: result };
        }
        console.log(`Returning result ${JSON.stringify(resp)}`);
        return resp;
    }
}
exports.HttpHandler = HttpHandler;
